packages/matter.js/src/cluster/server/AttributeServer.ts

```diff
@@ -64,21 +75,73 @@ export class FixedAttributeServer<T> extends BaseAttributeServer<T> {
         if (version !== undefined) {
             throw new Error("Version is not supported on fixed attributes");
         }
-        this.validator(value, this.name);
+        this.schema.validate(value);
         this.value = value;
     }
 }
 
+/**
+ * Base class for attribute servers that back mutable (non-fixed) attributes.
+ */
+export abstract class MutableAttributeServer<T> extends BaseAttributeServer<T> {
+    set(value: T, session?: Session<MatterDevice>) {
+        // TODO - ACL enforcement
+
+        this.setFabricIndex(value, session?.getAccessingFabric().fabricIndex, this.schema);
+
+        this.setRemote(value, session);
+    }
+
+    /**
+     * Set an attribute value in the context of a remote request.
+     */
+    abstract setRemote(value: T, session?: Session<MatterDevice>): void;
+
+    /**
+     * Set the value for the global "FabricIndex" attribute.
+     */
+    private setFabricIndex(value: any, fabricIndex: FabricIndex | undefined, schema: TlvSchema<any>) {
+        if (schema instanceof ArraySchema) {
+            if (Array.isArray(value)) {
+                for (const item of value) {
+                    this.setFabricIndex(item, fabricIndex, schema.elementSchema);
+                }
+            }
+        } else if (schema instanceof NullableSchema) {
+            this.setFabricIndex(value, fabricIndex, schema.schema);
+        } else if (schema instanceof ObjectSchema) {
+            if (typeof schema !== "object") {
+                return;
+            }
+
+            for (const k in schema.fieldDefinitions) {
+                const field = schema.fieldDefinitions[k] as FieldType<any>;
+
+                // 0xfe = global FieldIndex attribute
+                if (field.id === Globals.FabricIndex.id) {
+                    if (fabricIndex === undefined) {
+                        throw new FabricScopeError("Session is required to initialize fabricIndex attribute");
+                    }
+
+                    value[k] = fabricIndex;
+                } else {
+                    this.setFabricIndex(value[k], fabricIndex, field.schema);
+                }
+            }
+        }
+    }
+}
+
 /**
  * Attribute server for normal attributes that can be read and written.
  */
-export class AttributeServer<T> extends BaseAttributeServer<T> {
+export class AttributeServer<T> extends MutableAttributeServer<T> {
```

Comment:
Hm ... hsould not all fabric scoped data structs be contained within FabricScoped attributes?  Then this should be only for the Fabric scioped class? Then maybe even no extra class needed

Reply:
I partially created the intermediate class so ACL checks could be in once place.  I didn't see in the spec where they specify fabric scoped fields can't appear under non-scoped attributes but you might be right there

Reply:
Ok